crates/blockchain/blockchain.rs

```diff
@@ -206,5 +307,94 @@ fn get_total_blob_gas(tx: &EIP4844Transaction) -> u64 {
     GAS_PER_BLOB * tx.blob_versioned_hashes.len() as u64
 }
 
+fn total_difficulty_check<'a>(
+    head_block_hash: &'a H256,
+    head_block: &'a BlockHeader,
+    storage: &'a Store,
+) -> Result<(), InvalidForkChoice> {
+    if !head_block.difficulty.is_zero() || head_block.number == 0 {
+        let total_difficulty = storage
+            .get_block_total_difficulty(*head_block_hash)
+            .map_err(wrap)?;
+        let parent_total_difficulty = storage
+            .get_block_total_difficulty(head_block.parent_hash)
+            .map_err(wrap)?;
+        let terminal_total_difficulty = storage
+            .get_chain_config()
+            .map_err(wrap)?
+            .terminal_total_difficulty;
+        if terminal_total_difficulty.is_none()
+            || total_difficulty.is_none()
+            || head_block.number > 0 && parent_total_difficulty.is_none()
+        {
+            return Err(InvalidForkChoice::StoreError(StoreError::Custom(
+                "Total difficulties unavailable for terminal total difficulty check".to_string(),
+            )));
+        }
+        if total_difficulty.unwrap() < terminal_total_difficulty.unwrap().into() {
+            return Err(InvalidForkChoice::PreMergeBlock);
+        }
+        if head_block.number > 0
+            && parent_total_difficulty.unwrap() >= terminal_total_difficulty.unwrap().into()
+        {
+            return Err(InvalidForkChoice::StoreError(StoreError::Custom(
+                "Parent block is already post terminal total difficulty".to_string(),
+            )));
+        }
+    }
+    Ok(())
+}
+
+// Find branch of the blockchain connecting a block with the canonical chain. Returns the
+// number-hash pairs representing all blocks in that brunch. If genesis is reached and the link
+// hasn't been found, an error is returned.
+//
+// Return values:
+// - Err(StoreError): a db-related error happened.
+// - Ok(None): The block is not connected to the canonical chain.
+// - Ok(Some([])): the block is already canonical.
+// - Ok(Some(branch)): the "branch" is a sequence of blocks that connects the ancestor and the
+//   descendant.
+fn find_link_with_canonical_chain(
+    store: &Store,
+    block: &BlockHeader,
+) -> Result<Option<Vec<(BlockNumber, BlockHash)>>, StoreError> {
+    let mut block_number = block.number;
+    let block_hash = block.compute_block_hash();
+    let mut header = block.clone();
+    let mut branch = Vec::new();
+
+    if is_canonical(store, block_number, block_hash)? {
+        return Ok(Some(branch));
+    }
+
+    let Some(genesis_number) = store.get_earliest_block_number()? else {
+        return Err(StoreError::Custom(
+            "Earliest block number not found. Node setup must have been faulty.".to_string(),
```

Comment:
`store.get_earliest_block_number()` should not return an `Option`. There's always an earliest block (potentially genesis). If there is not, we should panic

Reply:
(maybe out of scope of this PR)